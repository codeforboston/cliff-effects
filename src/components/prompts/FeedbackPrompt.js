import React from 'react';
import { Button, Modal } from 'semantic-ui-react';

/**
 * Called with result of user interaction with on leave modal.
 * Receives true if the user chose to leave,
 * or false if the user chose to stay.
 *
 * @callback onLeaveCallback
 * @param ok {boolean} - Result of result of user interaction.
 */

/**
 * Modal with three options: stay, leave, or submit feedback;
 *
 * @param props {object}
 * @param props.callback {onLeaveCallback}
 * @param props.message {string}
 * @param props.open {boolean} - Whether the modal is visible.
 * @param props.isBlocking {boolean} - A secondary flag to control modal visibility.
 * @param props.leaveText {string}
 * @param props.stayText {string}
 */
class FeedbackPrompt extends React.Component {
  leave = (event) => {
    event.preventDefault();
    this.props.callback(true);
  };
  stay = (event) => {
    event.preventDefault();
    this.props.callback(false);
  };

  render() {
    const {
      isBlocking,
      header,
      message,
      open,
      leaveText,
      stayText,
      openFeedback,
    } = this.props;

    // If the user hasn't interacted with the form at all
    if (!isBlocking) {
      // just go ahead and leave without displaying prompt
      return null;
    }

    // Otherwise, set up the prompt
    let realLeave   = leaveText || 'Leave',
        realStay    = stayText || `Cancel`,
        realMessage = message;
    if (message === 'default') {
      realMessage = `Clicking "${realLeave}" will erase the information you have put into the form. Do you want to tell us something about the app first? That information could help us. You will still be able to come back and click "${realLeave}" afterwards.`;
    }

    return (
      <Modal
        mountNode = { document.getElementById(`App`) }
        open={ open }>
        <Modal.Header>{header || `Do you want to tell us more before you click "${realLeave}"?`}</Modal.Header>
        <Modal.Content>
          <p> {realMessage} </p>
        </Modal.Content>
        <Modal.Actions>
          <Button onClick={ this.stay }>{realStay}</Button>
          <Button onClick={ this.leave }>{realLeave}</Button>
          <Button
            onClick={ openFeedback }
            primary>Tell Us More
          </Button>
        </Modal.Actions>
      </Modal>
    );
  }
}

export default FeedbackPrompt;
