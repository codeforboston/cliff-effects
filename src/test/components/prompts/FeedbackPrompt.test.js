import React from 'react';
import { shallow } from 'enzyme';


import FeedbackPrompt from '../../../components/prompts/FeedbackPrompt';

test('A feedback prompt should match the snapshot', () => {
  let isBlocking = true;

  let promptData = {
    header:    'Test prompt data',
    leaveText: 'Reset',
  };

  const wrapper = shallow(
    <FeedbackPrompt
      { ...promptData }
      isBlocking={ isBlocking }
      openFeedback={ jest.fn() } />
  );
  expect(wrapper).toMatchSnapshot();
});
