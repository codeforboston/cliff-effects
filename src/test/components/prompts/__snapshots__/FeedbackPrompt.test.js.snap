// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A feedback prompt should match the snapshot 1`] = `
<Modal
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
>
  <ModalHeader>
    Test prompt data
  </ModalHeader>
  <ModalContent>
    <p>
       
       
    </p>
  </ModalContent>
  <ModalActions>
    <Button
      as="button"
      onClick={[Function]}
    >
      Reset
    </Button>
    <Button
      as="button"
      onClick={[Function]}
    >
      Stay
    </Button>
    <Button
      as="button"
      onClick={[Function]}
      primary={true}
    >
      Submit Feedback
    </Button>
  </ModalActions>
</Modal>
`;
